app/assets/javascripts/discourse/app/instance-initializers/set-custom-ajax-request-headers.js

```diff
@@ -0,0 +1,19 @@
+import { setCustomHeader } from "discourse/lib/ajax";
+import PreloadStore from "discourse/lib/preload-store";
+
+export default {
+  initialize() {
+    const customAjaxRequestHeaders = PreloadStore.get(
+      "customAjaxRequestHeaders"
+    );
+
+    if (
+      customAjaxRequestHeaders &&
+      Object.keys(customAjaxRequestHeaders).length > 0
+    ) {
+      for (const [header, value] of Object.entries(customAjaxRequestHeaders)) {
+        setCustomHeader(header, value);
```

Comment:
One complication here is that not all our requests go through the `ajax()` library. e.g. there are a few places (e.g.  presence, click tracking) where we use 'beacon' requests to ensure they are not aborted prematurely as the user navigates away.

https://github.com/discourse/discourse/blob/0f5d5403ea15fcb37b78f2b0e1057008ec1fa1a6/app/assets/javascripts/discourse/app/services/presence.js#L490-L490

So... perhaps we have to add some custom logic to each of those places? Or maybe we create a `sendBeacon` method in the `ajax` library so that it's centralised and can respect `setCustomHeader` 🤔  (that's assuming beacon requests even allow custom headers... I'm not sure)

Reply:
An idea: how about using a cookie instead of a magic header? That way it could be 100% server-side and it'll work for every type of request made in the browser session.

Something like

```ruby
ApplicationController.before_action do
  cookies["rspec_current_example_location"] ||= BlockRequestsMiddleware.current_example_location
end
```
and then the middleware would be something along the lines of

```ruby
cookie_location = request.cookies["rspec_current_example_location"]
if cookie_location != env["HTTP_RSPEC_CURRENT_EXAMPLE_LOCATION"]
  # block
end
```